p2p/simulations/events_test.go

```diff
@@ -0,0 +1,170 @@
+package simulations
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+)
+
+func init() {
+	log.Root().SetHandler(log.LvlFilterHandler(log.LvlInfo, log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
+}
+
+func TestNodeUpAndConn(t *testing.T) {
+
+	var quitC []chan struct{}
+	defer func() {
+		for _, q := range quitC {
+			if q != nil {
+				close(q)
+			}
+		}
+	}()
+
+	// create simulation network with 20 testService nodes
+	adapter := adapters.NewSimAdapter(adapters.Services{
+		"test": newTestService,
+	})
+	network := NewNetwork(adapter, &NetworkConfig{
+		DefaultService: "test",
+	})
+	nodeCount := 3
+	ids := make([]discover.NodeID, nodeCount)
+
+	for i := 0; i < nodeCount; i++ {
+		conf := adapters.RandomNodeConfig()
+		node, err := network.NewNodeWithConfig(conf)
+		ids[i] = node.ID()
+		if err != nil {
+			t.Fatalf("error creating node: %s", err)
+		}
+	}
+
+	trigger := make(chan discover.NodeID)
+	events := make(chan *Event)
+	sub := network.Events().Subscribe(events)
+	defer sub.Unsubscribe()
+	quitC = append(quitC, make(chan struct{}))
+
+	action := func(ctx context.Context) error {
+		go func(quitC chan struct{}) {
+			for {
+				select {
+				case ev := <-events:
+					if ev == nil {
+						panic("got nil event")
+					}
+					if ev.Type == EventTypeNode {
+						if ev.Node.Up {
+							log.Info("got node up event", "event", ev, "node", ev.Node.Config.ID)
+							trigger <- ev.Node.Config.ID
+						}
+					}
+
+				case <-quitC:
+					log.Warn("got quit action 1")
+					return
+				}
+
+			}
+		}(quitC[len(quitC)-1])
+		go func() {
+			for _, n := range ids {
+				if err := network.Start(n); err != nil {
+					t.Fatalf("error starting node: %s", err)
+				}
+				log.Info("network start returned", "node", n)
+			}
+		}()
+		return nil
+	}
+
+	check := func(ctx context.Context, nodeId discover.NodeID) (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+		}
+		log.Info("trigger expect up", "node", nodeId)
+		return true, nil
+	}
+
+	timeout := 30 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	result := NewSimulation(network).Run(ctx, &Step{
+		Action:  action,
+		Trigger: trigger,
+		Expect: &Expectation{
+			Nodes: ids,
+			Check: check,
+		},
+	})
+	if result.Error != nil {
+		t.Fatalf("simulation failed: %s", result.Error)
+	}
+	close(quitC[len(quitC)-1])
+	quitC[len(quitC)-1] = make(chan struct{})
+
+	action = func(ctx context.Context) error {
+		go func(quitC chan struct{}) {
+			for {
+				select {
+				case ev := <-events:
+					if ev.Type == EventTypeConn {
```

Comment:
check for `ev == nil` here too?

Reply:
I think more correct is to add `defer net.Shutdown()`, and move the `defer` function closing `quitC` after it. In that case, `ev` would never be `nil`